File: src/main/rs/emulate/legacy/widget/type/ContainerWidget.java
package rs.emulate.legacy.widget.type;

import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.OptionalInt;

import rs.emulate.legacy.widget.WidgetGroup;
import rs.emulate.legacy.widget.WidgetOption;
import rs.emulate.legacy.widget.script.LegacyClientScript;
import rs.emulate.shared.util.DataBuffer;
import rs.emulate.shared.util.Point;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;

/**
 * Contains properties used by {@link WidgetGroup#CONTAINER} {@link Widget}s.
 *
 * @author Major
 */
public final class ContainerWidget extends Widget {

	/**
	 * The {@link Point}s of the child Widgets.
	 */
	private List<Point> childPoints;

	/**
	 * The child ids of the Widget.
	 */
	private int[] children;

	/**
	 * Indicates whether or not the Widget is hidden.
	 */
	private boolean hidden;

	/**
	 * The scroll limit of the container Widget.
	 */
	private int scrollLimit;

	/**
	 * Creates the ContainerWidget.
	 * 
	 * @param id The id of the ContainerWidget.
	 * @param parent The parent id of the ContainerWidget, as an {@link OptionalInt}. Must not be {@code null}.
	 * @param optionType The {@link WidgetOption} of the ContainerWidget. Must not be {@code null}.
	 * @param content The content type of the ContainerWidget.
	 * @param width The width of the ContainerWidget, in pixels.
	 * @param height The width of the ContainerWidget, in pixels.
	 * @param alpha The alpha of the ContainerWidget, in pixels.
	 * @param hover The hover id of the ContainerWidget, as an {@link OptionalInt}. Must not be {@code null}.
	 * @param scripts The {@link List} of {@link LegacyClientScript}s. Must not be {@code null}.
	 * @param option The {@link Option} of the ContainerWidget, as an {@link Optional}. Must not be {@code null}.
	 * @param hoverText The hover text of the ContainerWidget, as an {@link Optional}. Must not be {@code null}.
	 * @param scrollLimit The scroll limit of the container ContainerWidget.
	 * @param hidden Whether or not the ContainerWidget is hidden.
	 * @param children The child ids of the ContainerWidget.
	 * @param childPoints The {@link Point}s of the child ContainerWidgets.
	 */
	public ContainerWidget(int id, OptionalInt parent, WidgetOption optionType, int content, int width, int height, int alpha,
			OptionalInt hover, List<LegacyClientScript> scripts, Optional<Option> option, Optional<String> hoverText,
			int scrollLimit, boolean hidden, int[] children, List<Point> childPoints) {
		super(id, parent, WidgetGroup.CONTAINER, optionType, content, width, height, alpha, hover, scripts, option, hoverText);

		this.scrollLimit = scrollLimit;
		this.hidden = hidden;
		this.children = Objects.requireNonNull(children, "Child array must not be null.").clone();
		this.childPoints = ImmutableList.copyOf(Objects.requireNonNull(childPoints, "Child Point List must not be null."));

		Preconditions.checkArgument(children.length == childPoints.size(), "Child ids and child points must be of equal length.");
	}

	@Override
	public DataBuffer encodeBespoke() {
		DataBuffer buffer = DataBuffer.allocate((2 + children.length) * Short.BYTES + Byte.BYTES);
		buffer.putShort(scrollLimit);
		buffer.putBoolean(hidden);

		int size = children.length;
		buffer.putShort(size);

		for (int index = 0; index < size; index++) {
			buffer.putShort(children[index]);

			Point point = childPoints.get(index);
			buffer.putShort(point.getX());
			buffer.putShort(point.getY());
		}

		return buffer.flip().asReadOnlyBuffer();
	}

}

File: legacy/src/main/kotlin/rs/emulate/legacy/widget/type/ContainerWidget.kt
package rs.emulate.legacy.widget.type

import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import rs.emulate.legacy.widget.Widget
import rs.emulate.legacy.widget.WidgetGroup
import rs.emulate.legacy.widget.WidgetGroup.CONTAINER
import rs.emulate.legacy.widget.WidgetOption
import rs.emulate.legacy.widget.script.LegacyClientScript
import rs.emulate.util.Point

/**
 * Contains properties used by [WidgetGroup.CONTAINER] [Widget]s.
 *
 * @param id The id of the ContainerWidget.
 * @param parent The parent id of the ContainerWidget.
 * @param optionType The [WidgetOption] of the ContainerWidget.
 * @param content The content type of the ContainerWidget.
 * @param width The width of the ContainerWidget, in pixels.
 * @param height The width of the ContainerWidget, in pixels.
 * @param alpha The alpha of the ContainerWidget, in pixels.
 * @param hover The hover id of the ContainerWidget.
 * @param scripts The [List] of [LegacyClientScript]s.
 * @param option The [Option] of the ContainerWidget.
 * @param hoverText The hover text of the ContainerWidget.
 * @param scrollLimit The scroll limit of the container ContainerWidget.
 * @param hidden Whether or not the ContainerWidget is hidden.
 * @param children The child ids of the ContainerWidget.
 * @param childPoints The [Point]s of the child ContainerWidgets.
 */
class ContainerWidget( // TODO this constructor size is PTSD-inducing
    id: Int, parent: Int?, optionType: WidgetOption, content: Int, width: Int, height: Int, alpha: Int,
    hover: Int?, scripts: List<LegacyClientScript>, option: Option?, hoverText: String?,
    private val scrollLimit: Int,
    private val hidden: Boolean,
    children: IntArray,
    childPoints: List<Point>
) : Widget(id, parent, CONTAINER, optionType, content, width, height, alpha, hover, scripts, option, hoverText) {

    /**
     * The [Point]s of the child Widgets.
     */
    private val childPoints: List<Point> = childPoints.toList()

    /**
     * The child ids of the Widget.
     */
    private val children: IntArray = children.clone()

    init {
        require(children.size == childPoints.size) { "Child ids and child points must be of equal length." }
    }

    public override fun encodeBespoke(): ByteBuf {
        val buffer = Unpooled.buffer((2 + children.size) * java.lang.Short.BYTES + java.lang.Byte.BYTES)
        buffer.writeShort(scrollLimit)
        buffer.writeBoolean(hidden)

        val size = children.size
        buffer.writeShort(size)

        for (index in 0 until size) {
            buffer.writeShort(children[index])

            val point = childPoints[index]
            buffer.writeShort(point.x)
            buffer.writeShort(point.y)
        }

        return buffer
    }

}
